### DISCUSSION TITLE
Which version of Opengl am I getting in Openglcontrolbase and which shader version is best?

### DISCUSSION THREAD

**Discussion by ejongejans** [AvaloniaUI/Avalonia#15055] [Q&A] [answered]
Hi, 

I'm currently using OpenGLControlbase + WGL=true + Silk.NET on windows. 
However it seems at some users this leads to unstable situations and weird bugs. 
I read a lot about WGL being the cause of that; so wanted to switch to Angle.

When I turn WGL=false i get opengl 4.0 over Angle.
However i have various issues in my shaders. 

_' : invalid version directive
ERROR: 0:6: 'layout' : syntax error
ERROR: 0:4: 'in' : storage qualifier supported in GLSL ES 3.00 and above only
ERROR: 0:4: '' : No precision specified for (float)
ERROR: 0:6: 'flat' : Illegal use of reserved word_

- How to know which OpenGL version you will receive? 
- I thought angle was just made for OpenGL ES 2/3, but it's not instantiated by default? Is there a way to do this? 
- Any hint on which shader version to use best? I see in the opengl examples the code suggests 100, 120 or 150, and in the silk example 330 core is used. 

I don't need a lot of specific functionality; openGL es 3 should do fine, or opengl 4. 

kind regards, 
Elco

**Comment by kekekeks** [comment]
You can query the version using `gl.ContextInfo.Version` property. The version you get is platform dependent.

As of shaders, it's recommended to use https://github.com/KhronosGroup/SPIRV-Cross to transpile your shaders to various formats/versions.

**Reply by ejongejans** [reply]
Is there anything to say about what different flavours one can expect? 

i.e. 
Opengl 4 on windows?
Opengl ES 3.0 on mac/linux ? 

Is this determined somewhere in Avalonia? I try to understand where the context gets created but I can't seem to find a clear place. 
Or is the underlying Angle implementation deciding what is there and picks the best one?

**Reply by ejongejans** [reply]
On my work PC (geforce card) I get OpenGL 4. 
On my home PC (intel graphics onboard) i get OpenGL ES 3.0.0 (ANGLE 2.1.0.b94851945092)

What determines what gets selected?

**Reply by ejongejans** [reply]
My goal is; if I know which different flavours there are roughly; I can write some simplistic shader code that either works on all or has some simple if/else statements to fix any strange difference.

**Reply by maxkatz6** [reply]
Specifically on Windows, you should by default expect ES 3.0. As that's what Angle supports - https://github.com/google/angle?tab=readme-ov-file#level-of-opengl-es-support-via-backing-renderers

You can change Win32PlatformOptions.RenderingMode to WGL in order to supports native GL - https://github.com/AvaloniaUI/Avalonia/blob/master/src/Windows/Avalonia.Win32/Win32PlatformOptions.cs#L114
But it's not recommended, as Angle normally has better stability and performance. Though depends on your needs.

**Reply by maxkatz6** [reply, marked as answer]
On other platforms it pretty much depends on what OS can provide. Like, Linux will normally use GLX, but you can set allowed profile versions if needed: https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.X11/X11Platform.cs#L318-L326

**Reply by ejongejans** [reply]
Thanks! 

I'm getting OpenGL es 30 on all my windows PC's now. Rewrote all my shaders and it seems to work.
